packages/react-reconciler/src/__tests__/ReactSuspenseCallback-test.internal.js

```diff
@@ -0,0 +1,209 @@
+    ReactFeatureFlags = require('shared/ReactFeatureFlags');
+    ReactFeatureFlags.enableSuspenseCallback = true;
+
+    React = require('react');
+    ReactNoop = require('react-noop-renderer');
+    Scheduler = require('scheduler');
+  });
+
+  function text(t) {
+    return {text: t, hidden: false};
+  }
+
+  function createThenable() {
+    let completed = false;
+    const resolveRef = {current: null};
+    let promise = {
+      then(resolve, reject) {
+        resolveRef.current = () => {
+          completed = true;
+          resolve();
+        };
+      },
+    };
+
+    const PromiseComp = () => {
+      if (!completed) {
+        throw promise;
+      }
+      return 'Done';
+    };
+    return {promise, resolveRef, PromiseComp};
+  }
+
+  it('1 then 0 suspense callback', () => {
+    const {promise, resolveRef, PromiseComp} = createThenable();
+
+    let lastThenablesCallback = null;
+    const suspenseCallback = thenables => {
+      lastThenablesCallback = thenables;
+    };
+
+    const element = (
+      <React.Suspense suspenseCallback={suspenseCallback} fallback={'Waiting'}>
+        <PromiseComp />
+      </React.Suspense>
+    );
+
+    ReactNoop.render(element);
+    expect(Scheduler).toFlushWithoutYielding();
+    expect(ReactNoop.getChildren()).toEqual([text('Waiting')]);
+    expect(lastThenablesCallback).toEqual(new Set([promise]));
+
+    resolveRef.current();
+    expect(Scheduler).toFlushWithoutYielding();
+    expect(ReactNoop.getChildren()).toEqual([text('Done')]);
+    expect(lastThenablesCallback).toEqual(new Set([]));
+  });
+
+  it('2 then 1 then 0 suspense callback', () => {
+    const {
+      promise: promise1,
+      resolveRef: resolveRef1,
+      PromiseComp: PromiseComp1,
+    } = createThenable();
+    const {
+      promise: promise2,
+      resolveRef: resolveRef2,
+      PromiseComp: PromiseComp2,
+    } = createThenable();
+
+    let lastThenablesCallback1 = null;
+    const suspenseCallback1 = thenables => {
+      lastThenablesCallback1 = thenables;
+    };
+
+    const element = (
+      <React.Suspense
+        suspenseCallback={suspenseCallback1}
+        fallback={'Waiting Tier 1'}>
+        <PromiseComp1 />
+        <PromiseComp2 />
+      </React.Suspense>
+    );
+
+    ReactNoop.render(element);
+    expect(Scheduler).toFlushWithoutYielding();
+    expect(ReactNoop.getChildren()).toEqual([text('Waiting Tier 1')]);
+    expect(lastThenablesCallback1).toEqual(new Set([promise1, promise2]));
+
+    resolveRef1.current();
+    ReactNoop.render(element);
+    expect(Scheduler).toFlushWithoutYielding();
+    expect(ReactNoop.getChildren()).toEqual([text('Waiting Tier 1')]);
+    expect(lastThenablesCallback1).toEqual(new Set([promise2]));
+
+    resolveRef2.current();
+    ReactNoop.render(element);
+    expect(Scheduler).toFlushWithoutYielding();
+    expect(ReactNoop.getChildren()).toEqual([text('Done'), text('Done')]);
+    expect(lastThenablesCallback1).toEqual(new Set([]));
```

Comment:
It would be good to assert on the number of callbacks. One pattern that we use a lot is:
```
let ops = [];
const suspenseCallback1 = thenables => {
  ops.push(thenables);
};
...
expect(ops).toEqual([new Set([promise2])]);
ops = [];
...
expect(ops).toEqual([]);
```

That way you're asserting on the number of calls and get to see the wrong sequence in the logs.

I was originally thrown off by the error because I thought the persistent renderer was actually passing the wrong thenables here but now I realize that it just wasn't invoking it.

Reply:
I was afraid of over-specifying my test. As-is I don't diff the thenables set before invoking the callback. So unrelated updates to the suspense component may cause more or less callback, which I think is related to the failure we got.

Should I ten diff the thenables? In the complete or commit phase?